### PR TITLE
[JENKINS-34438] Activate JSHint during the build process

### PR DESCRIPTION
**Pull request by recena**
[JENKINS-34438](https://issues.jenkins-ci.org/browse/JENKINS-34438)

Motivation: avoiding these fixes https://github.com/jenkinsci/jenkins/pull/2249

```
[INFO] --- frontend-maven-plugin:1.0:gulp (gulp bundle) @ jenkins-war ---
[INFO] Running 'gulp.js bundle lint' in /Users/recena/Development/projects/jenkins/war
[INFO] [19:59:36] Maven project
[INFO] [19:59:36] Setting defaults
[INFO] [19:59:36] Bundle will be generated in directory 'src/main/webapp/jsbundles' as 'page-init.js'.
[INFO] [19:59:36] Bundle will be generated in directory 'src/main/webapp/jsbundles' as 'pluginSetupWizard.js'.
[INFO] [19:59:36] Bundle will be generated in directory 'src/main/webapp/jsbundles' as 'config-tabbar.js'.
[INFO] [19:59:36] Bundle will be generated in directory 'src/main/webapp/jsbundles' as 'config-scrollspy.js'.
[INFO] [19:59:36] Bundle will be generated in directory 'src/main/webapp/jsbundles' as 'add-item.js'.
[INFO] [19:59:36] Using gulpfile ~/Development/projects/jenkins/war/gulpfile.js
[INFO] [19:59:36] Starting 'log-env'...
[INFO] [19:59:36] Source Dirs:
[INFO] [19:59:36]  - src: src/main/js,src/main/less
[INFO] [19:59:36]  - test: src/test/js
[INFO] [19:59:36] Finished 'log-env' after 727 μs
[INFO] [19:59:36] Starting 'bundle_page-init'...
[INFO] [19:59:36] Starting 'bundle_pluginSetupWizard'...
[INFO] [19:59:37] LESS CSS pre-processing completed to 'src/main/webapp/jsbundles'.
[INFO] [19:59:37] Starting 'bundle_config-tabbar'...
[INFO] [19:59:37] LESS CSS pre-processing completed to 'src/main/webapp/jsbundles'.
[INFO] [19:59:37] Starting 'bundle_config-scrollspy'...
[INFO] [19:59:37] LESS CSS pre-processing completed to 'src/main/webapp/jsbundles'.
[INFO] [19:59:37] Starting 'bundle_add-item'...
[INFO] [19:59:37] LESS CSS pre-processing completed to 'src/main/webapp/jsbundles'.
[INFO] [19:59:37] Starting 'lint'...
[INFO] src/main/js/widgets/config/tabbar.js: line 42, col 78, Missing semicolon.
[INFO] 
[INFO] 1 error
[INFO] [19:59:38] Finished 'lint' after 1.83 s
[INFO] [19:59:39] Finished 'bundle_page-init' after 2.99 s
[INFO] [19:59:39] Finished 'bundle_add-item' after 2.76 s
[INFO] [19:59:39] Finished 'bundle_config-tabbar' after 2.82 s
[INFO] [19:59:39] Finished 'bundle_config-scrollspy' after 2.82 s
[INFO] [19:59:40] Finished 'bundle_pluginSetupWizard' after 3.06 s
[INFO] [19:59:40] Starting 'bundle'...
[INFO] [19:59:40] Finished 'bundle' after 13 μs
```

@reviewbybees esp. @tfennelly and @kzantow

### CODE REVIEW

**Comment by recena** [comment]
It seems this configuration does not  make the build fail....

File: war/pom.xml [hunk 1]
```diff
@@ -733,7 +733,7 @@ THE SOFTWARE.
                                 <goal>gulp</goal>
                             </goals>
                             <configuration>
-                                <arguments>bundle</arguments>
+                                <arguments>bundle lint</arguments>
```
Comment (kzantow) [inline war/pom.xml]: probably should do: lint bundle

**Comment by recena** [comment]
```
[INFO] [20:08:30] Starting 'lint'...
[INFO] src/main/js/widgets/config/tabbar.js: line 43, col 77, Missing semicolon.
[INFO] 
[INFO] 1 error
[INFO] [20:08:31] 'lint' errored after 1.73 s
[INFO] [20:08:31] Error in plugin 'gulp-jshint'
[INFO] Message:
[INFO]     JSHint failed for: src/main/js/widgets/config/tabbar.js
[INFO] [20:08:32] Finished 'bundle_page-init' after 2.94 s
[INFO] [20:08:32] Finished 'bundle_add-item' after 2.73 s
[INFO] [20:08:33] Finished 'bundle_config-tabbar' after 2.8 s
[INFO] [20:08:33] Finished 'bundle_config-scrollspy' after 2.79 s
[INFO] [20:08:33] Finished 'bundle_pluginSetupWizard' after 2.99 s
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Jenkins main module ................................ SUCCESS [  1.164 s]
[INFO] Jenkins cli ........................................ SUCCESS [  0.620 s]
[INFO] Jenkins core ....................................... SUCCESS [  3.756 s]
[INFO] Jenkins war ........................................ FAILURE [ 10.488 s]
[INFO] Tests for Jenkins core ............................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 17.098 s
[INFO] Finished at: 2016-04-25T20:08:33+02:00
[INFO] Final Memory: 27M/348M
```

File: war/gulpfile.js [hunk 2]
```diff
@@ -44,3 +47,12 @@ builder.bundle('src/main/js/add-item.js')
     .withExternalModuleMapping('jquery-detached', 'core-assets/jquery-detached:jquery2')
     .less('src/main/js/add-item.less')
     .inDir('src/main/webapp/jsbundles');
+
+//
+// JSHint, a JavaScript Code Quality Tool
+//
+gulp.task('lint', function() {
+    return gulp.src('./src/main/js/**/*.js')
```
Comment (kzantow) [inline war/gulpfile.js]: for consistency maybe do `.src('src/m...` without `./`

File: war/pom.xml (hunk 1 above)
Comment (kwhetstone) [inline war/pom.xml]: I definitely agree on the failure order.

Comment (recena) [inline war/pom.xml]: @kzantow Sure! Then we have to do the same here https://github.com/jenkinsci/plugin-pom/blob/master/pom.xml#L1102

**Comment by recena** [comment]
@kzantow Your comments were addressed here https://github.com/jenkinsci/jenkins/pull/2289/commits/fc4435c29fbc6afd1165d0da5e14ccfa27e05827

**Comment by kzantow** [comment]
:bee: nice!

**Comment by kwhetstone** [comment]
:bee:

**Comment by recena** [comment]
@reviewbybees done

**Comment by tfennelly** [comment]
@recena I'd have expected the `pom.xml` changes on their own to do the trick for you i.e. I think the "lint" task should already be defined.

**Comment by recena** [comment]
@tfennelly In my opinion, everything related with _frontend_ should be defined on `gulpfile.js` and invoke to **gulp** by default.

I'm not sure if I understood well. Could you explain it again?

**Comment by recena** [comment]
@daniel-beck I think this PR is ready to merge.

**Comment by recena** [comment]
@jenkinsci/code-reviewers your feedback is welcome!

**Comment by oleg-nenashev** [comment]
I'm not the best code reviewer for JS, but it needs a merge conflict fix at least

**Comment by recena** [comment]
@oleg-nenashev You comment was addressed in my last commit.

File: war/src/main/js/pluginSetupWizardGui.js [hunk 3]
```diff
@@ -913,7 +913,7 @@ var createPluginSetupWizard = function(appendTarget) {
 								if (j.correlationId) {
 									selectedPluginNames.push(j.name);
 								}
-								setFailureStatus(j)
+								setFailureStatus(j);
```
Comment (recena) [inline war/src/main/js/pluginSetupWizardGui.js]: Unrelated with this PR but needed.

File: war/src/main/js/api/pluginManager.js [hunk 4]
```diff
@@ -185,7 +185,7 @@ exports.installPluginsDone = function(handler) {
 			handler.call({ isError: true, message: errorThrown });
 		}
 	});
-}
+};
```
Comment (recena) [inline war/src/main/js/api/pluginManager.js]: Unrelated with this PR but needed.

File: war/src/main/js/pluginSetupWizardGui.js (hunk 3 above)
Comment (oleg-nenashev) [inline war/src/main/js/pluginSetupWizardGui.js]: Separate bug? What's the impact?

Comment (recena) [inline war/src/main/js/pluginSetupWizardGui.js]: @oleg-nenashev This PR has the goal to detect this kind of bugs. This bug was introduced recently (after my PR), we need to solve it but the build will fail.

**Comment by recena** [comment]
@daniel-beck @oleg-nenashev Thanks so much for your time

**Pull request merged**